xc/device.ml

```diff
@@ -1963,6 +1974,96 @@ module Dm = struct
         List.iter (write_local_domain "data/") ["updated"]
       | _ -> ()
 
+  (* Create two hash tables to store <domid, qmp connction> and <qmp connection, domid> *)
+  let domid_to_connection_tbl = Hashtbl.create 128
+  let connection_to_domid_tbl = Hashtbl.create 128
+
+  let qmp_event_handle domid qmp_event =
+    (* This function will be extended to handle qmp events *)
+    debug "Get QMP event, domain-%d: %s" domid qmp_event.event
+ 
+  let qmp_event_thread () =
+    let epfd = Epoll.create1 [Epoll.EPOLL_CLOEXEC] in
+    Epoll.ctl epfd Epoll.EPOLL_CTL_ADD qmp_event_pipe_r [Epoll.EPOLLIN];
+
+    let remove domid =
+      if Hashtbl.mem domid_to_connection_tbl domid
```

Comment:
I see now why you may want to find a fd given a domid passed to remove. 

I think you should abstract both hashtables in a single data structure that's easy to use. Something like this inside module QMP_Event:
```
module Lookup = 
  let ftod, dtof = Hashtbl.create 16, Hashtbl.create 16
  let add fd domid =
    Hashtbl.replace ftod fd domid;
    Hashtbl.replace dtof domid fd
  let remove fd domid =
    Hashtbl.remove ftod fd domid;
    Hashtbl.remove dtof domid fd
  let domid_of fd = try Some (Hashtbl.find ftod fd) with Not_found -> None
  let fd_of domid = try Some (Hashtbl.find dtof domid) with Not_found -> None
end
```

Reply:
Thanks, added the `Lookup` module, I also put `Epoll.ctl` and `Qmp_protocol.connect/close` into this module, I think it's straightforward.